diff -urN elf2flt-20070723.orig/elf2flt.c elf2flt-20070723/elf2flt.c
--- elf2flt-20070723.orig/elf2flt.c	Tue Nov 14 23:20:08 2006
+++ elf2flt-20070723/elf2flt.c	Tue Jul 24 13:33:18 2007
@@ -236,6 +236,10 @@
     return NULL;
 
   symbol_table = (asymbol **) malloc (storage_needed);
+  if (symbol_table == NULL) {
+      perror("malloc");
+      exit(1);
+  }
 
   number_of_symbols = bfd_canonicalize_symtab (abfd, symbol_table);
   
@@ -492,7 +496,12 @@
 	}
 
 	symb = get_symbols(rel_bfd, &nsymb);
-	relpp = (arelent **) xmalloc(relsize);
+	relpp = (arelent **) malloc(relsize);
+	if (relpp == NULL) {
+	  perror("malloc");
+	  exit(1);
+	}
+
 	relcount = bfd_canonicalize_reloc(rel_bfd, r, relpp, symb);
 	if (relcount <= 0) {
 		if (verbose)
@@ -1975,6 +1984,10 @@
   }
 
   text = malloc(text_len);
+  if (text == NULL) {
+      perror("malloc");
+      exit(1);
+  }
 
   if (verbose)
     printf("TEXT -> vma=0x%x len=0x%x\n", text_vma, text_len);
@@ -1995,6 +2008,10 @@
     exit (2);
   }
   data = malloc(data_len);
+  if (data == NULL) {
+      perror("malloc");
+      exit(1);
+  }
 
   if (verbose)
     printf("DATA -> vma=0x%x len=0x%x\n", data_vma, data_len);
@@ -2079,6 +2096,10 @@
   
   if (!ofile) {
     ofile = malloc(strlen(fname) + 5 + 1); /* 5 to add suffix */
+    if (ofile == NULL) {
+	perror("malloc");
+	exit(1);
+    }
     strcpy(ofile, fname);
     strcat(ofile, ".bflt");
   }
diff -urN elf2flt-20070723.orig/elf2flt.ld elf2flt-20070723/elf2flt.ld
--- elf2flt-20070723.orig/elf2flt.ld	Fri Jul 21 15:11:04 2006
+++ elf2flt-20070723/elf2flt.ld	Tue Jul 24 13:33:18 2007
@@ -51,9 +51,6 @@
 		. = ALIGN(0x20) ; 
 		LONG(-1)
 		. = ALIGN(0x20) ; 
-R_RODAT		*(.rodata)
-R_RODAT		*(.rodata1)
-R_RODAT		*(.rodata.*)
 R_RODAT		*(.gnu.linkonce.r*)
 		*(.data)
 		*(.data1)
@@ -62,7 +59,9 @@
 		*(.data1)
 		*(.eh_frame)
 		*(.gcc_except_table)
-
+R_RODAT		*(.rodata)
+R_RODAT		*(.rodata1)
+R_RODAT		*(.rodata.*)
 		/* Microblaze has .sdata and .sbss (small bss).  They must
 		   be contiguous, so please don't move any of this. JW */
 		_ssrw = . ;			
diff -urN elf2flt-20070723.orig/ld-elf2flt.in elf2flt-20070723/ld-elf2flt.in
--- elf2flt-20070723.orig/ld-elf2flt.in	Wed Sep 13 02:10:30 2006
+++ elf2flt-20070723/ld-elf2flt.in	Tue Jul 24 13:44:12 2007
@@ -157,7 +157,9 @@
 
 		if [ "@emit_relocs@" = "yes" ]
 		then
-		 $LINKER $EMUL $SDIRS -T $LDSCRIPT -q -o "$OFILE.gdb" $ARG1             ||exit $?
+		 $LINKER $EMUL -r -d -o "$OFILE.elf2flt"  $ARG1                          ||exit $?
+		 $LINKER $EMUL $SDIRS -T $LDSCRIPT -q -o "$OFILE.gdb" "$OFILE.elf2flt"   ||exit $?
+		 rm -f "$OFILE.elf2flt"
 		 RFILE="$OFILE.gdb"
 		 FLTFLAGS="$FLTFLAGS -a"
 		else
